tests/python_package_test/conftest.py

```diff
@@ -0,0 +1,9 @@
+# `test_cpu_and_gpu_work` needs to be executed before any of the dask tests
+# because the AMD OpenCL driver won't load on Ubuntu 14.04 otherwise.
+# We simply make it first here.
```

Comment:
Could you tell us more about this? What specifically happened and how did you narrow it down to `dask` tests being the problem?

If library-loading order affects whether or not this works, then I worry that `lightgbm`'s users might experience whatever issue you're working around with this change to the tests.

Linking a possibly-related conversation: https://github.com/microsoft/LightGBM/issues/5106#issuecomment-1125634723

Reply:
The issue is that the statically-linked OpenCL ICD loader is not able to load the AMD App SDK driver that we use for testing on the CPU. The call to `dlopen` fails with the error `cannot load any more object with static TLS`. This seems to only happen with the AMD App SDK. I have tested with an NVIDIA GPU and with the PoCL CPU implementation and it worked fine with both.

When running under Python 3.8 on Ubuntu 14.04, the AMD App SDK driver fails to load immediately. With Python 3.7, it only fails to load if the `dask` tests run first. I isolated that by running various combinations and orders of tests.

The whole issue seems to come from the AMD App SDK driver using statically allocated thread local storage (STATIC_TLS), which means that the amount of TLS objects is limited, and the library loading order can influence when the limit gets reached.

It's worth noting that both neither the NVIDIA OpenCL driver, nor the PoCL one use STATIC_TLS, so the wheel works fine with them under various versions of Python. I do not have access to the required hardware, but I have checked that the various ICDs available on Ubuntu 20.04 (Mesa, Beignet, and Intel) do not use STATIC_TLS either.

Reply:
I have tried running the tests with the `gpu` and `bdist` tasks with PoCL instead of the AMD App SDK and they all succeed, even the `dask` tests that are currently disabled with `gpu`. I think that using PoCL would make a lot of sense, as it is open source and actively maintained, as opposed to the AMD App SDK. It even enables testing of the Integrated OpenCL build on arm64!

I have a separate branch with these changes — if you're favourable to this idea, would you like me to raise another PR to get this done first, or would you like me to add it to this PR?

Reply:
@jgiannuzzi Thanks a lot for the detailed explanation!

> `cannot load any more object with static TLS`

Ah, I remember that annoying error! We've hit it already several times.
https://github.com/microsoft/LightGBM/pull/4893#issuecomment-1039204806
https://github.com/microsoft/LightGBM/pull/5022#issue-1146283008

> I think that using PoCL would make a lot of sense, as it is open source and actively maintained, as opposed to the AMD App SDK.

Absolutely agree with you!

> It even enables testing of the Integrated OpenCL build on arm64!

Awesome!

> I have a separate branch with these changes — if you're favourable to this idea, would you like me to raise another PR to get this done first, or would you like me to add it to this PR?

We'll really appreciate PR with those changes. I believe we can have it prior to this PR because due to low maintenance activity of this project we prefer small focused of one feature/fix self-contained PRs. Otherwise, big PRs can have a long review cycle.

Reply:
Great, I opened #5282!

Reply:
We install AMD OpenCL for Linux here
https://github.com/microsoft/LightGBM/blob/11110c540eb84f519da299ea9eb76cc23a0ca570/.ci/setup.sh#L68-L82
and for Windows here
https://github.com/microsoft/LightGBM/blob/master/.ci/install_opencl.ps1
https://github.com/microsoft/LightGBM/blob/11110c540eb84f519da299ea9eb76cc23a0ca570/.ci/test_windows.ps1#L77-L78

Reply:
> Great, I opened https://github.com/microsoft/LightGBM/pull/5282!

Thanks a lot!